### ISSUE TITLE
[typing] Ok to add typing_extensions as dep for older Pythons?

### ISSUE THREAD

**Issue by madig** [fonttools/fonttools#2163]
I'm looking at adding typing to some parts of the code base (currently: pens) to make them nicer to work with in IDEs and mypy.

Example: looking at `AbstractPointPen.addPoint`, the `segmentType` can be one of a few strings or `None`. Python's `typing.Literal` covers that, but was added only in Python 3.8. For older Pythons, there's a backport package at https://pypi.org/project/typing-extensions/ that contains that type.

Another example is `Protocol`, which can be used to declare that a class implements some interface, but without necessarily deriving from some common base class. I suppose pens fit that description, but I'm not sure if there is some `isinstance` shenanigans going on somewhere outside fontTools?

Long story short, is it ok to add `typing-extensions` as a conditional dependency for Python < 3.8?

**Comment by justvanrossum** [comment]
Shouldn't `AbstractPen` and `AbstractPointPen` become proper "abc"s first?

1. I'm personally not very keen on adding typing everywhere: it's a huge amount of additional noise for a remarkably small benefit. There's a lot of overengineering here.

2. To add another dependency (even if only for < 3.9) to get the ultimate precision for a call signature in a single method seems overkill.

**Comment by madig** [comment]
I can do the ABC stuff, yeah!

1. Typing is indeed some work, but I found the niceness adding up. A library with typing is nicer to use than one without in my experience.
2. Oh well :o

**Comment by madig** [comment]
https://www.youtube.com/watch?v=Lj_9TyT3V98

Here someone from Facebook talks about gradually adding types to production code.

**Comment by behdad** [comment]
> Long story short, is it ok to add `typing-extensions` as a conditional dependency for Python < 3.8?

Or just requirement to 3.8? What is the timeline for that?

**Comment by anthrotype** [comment]
> What is the timeline for that?

3.7 is EOL in end of June 2023, but I've already seen a few projects dropping support for it.
https://endoflife.date/python

**Comment by behdad** [comment]
Close?

**Comment by madig** [comment]
Yes.
